File: 2026/2026-06-17-cloudflare-workers-vectorize-agent/python-cloudflare/src/renderer.py
import json

from pyodide.ffi import create_proxy


class SseStreamRenderer:
    """Writes agent events to a JavaScript stream as soon as they happen."""

    def __init__(self, writer):
        self.writer = writer
        self.closed = False

    async def emit(self, event: str, payload: dict):
        """Writes one typed SSE event to the response stream."""

        if self.closed:
            return

        chunk = f"event: {event}\ndata: {json.dumps(payload)}\n\n".encode()
        try:
            await write_bytes(self.writer, chunk)
        except BaseException:
            self.closed = True

    async def close(self):
        """Closes the streaming response writer."""

        if self.closed:
            return

        self.closed = True
        try:
            await self.writer.close()
        except BaseException:
            pass


async def write_bytes(writer, body: bytes):
    """Writes Python bytes into a JavaScript WritableStream writer."""

    proxy = create_proxy(body)
    buffer = proxy.getBuffer()
    try:
        await writer.write(buffer.data.slice())
    finally:
        buffer.release()
        proxy.destroy()

File: 2026/2026-06-17-cloudflare-workers-vectorize-agent/full-stack-cloudflare/apps/agent-worker/src/renderer.ts
export type AgentEvent =
  | "status"
  | "iteration"
  | "reasoning"
  | "model_done"
  | "tool_call"
  | "tool_result"
  | "token"
  | "done"
  | "error";

export interface AgentRenderer {
  handleEvent(event: AgentEvent, payload: Record<string, unknown>): Promise<void> | void;
}

/**
 * Converts agent events into SSE chunks for `/ask/stream`.
 * Used by `app.ts` when creating a streaming response.
 */
export class SseRenderer implements AgentRenderer {
  private readonly encoder = new TextEncoder();
  private closed = false;

  /**
   * Stores the response stream controller for later SSE writes.
   * Constructed by `streamAsk` in `app.ts`.
   */
  constructor(private readonly controller: ReadableStreamDefaultController<Uint8Array>) {}

  /**
   * Emits one typed SSE event.
   * Called by `runAgent` through the `AgentRenderer` interface.
   */
  handleEvent(event: AgentEvent, payload: Record<string, unknown>) {
    if (this.closed) {
      return;
    }

    try {
      this.controller.enqueue(this.encoder.encode(`event: ${event}\ndata: ${JSON.stringify(payload)}\n\n`));
    } catch {
      this.closed = true;
    }
  }

  /**
   * Closes the SSE stream after agent work completes.
   * Called by `streamAsk`; ignores cases where the client disconnected first.
   */
  close() {
    if (this.closed) {
      return;
    }

    try {
      this.controller.close();
    } catch {
      // The browser or curl process may have already closed the connection.
    } finally {
      this.closed = true;
    }
  }
}
